tuna.go

```diff
@@ -777,21 +809,49 @@ func (c *Common) measureBandwidth(delayMeasuredSubs types.Nodes, n int) types.No
 				min, max, err := tunaUtil.BandwidthMeasurementClient(encryptedConn, int(c.MeasurementBytesDownLink), bandwidthTimeout)
 				dur := time.Since(timeStart)
 				if err != nil {
-					if _, ok := err.(net.Error); !ok {
-						log.Println(err)
+					if c.measureStorage != nil {
+						c.measureStorage.AddBadNode(sub.Metadata.Ip, &storage.BadNode{
+							IP:      sub.Metadata.Ip,
+							Address: sub.Address,
+						})
+						err = c.measureStorage.SaveBadNodes()
+						if err != nil {
+							log.Println(err)
+						}
+						log.Printf("add bad node: %s", sub.Metadata.Ip)
 					}
 					return
 				}
 
 				log.Printf("address: %s, bandwidth: %f - %f KB/s, time: %s", addr, min/1000, max/1000, dur)
 
+				if c.measureStorage != nil {
+					metadata, err := proto.Marshal(sub.Metadata)
+					if err != nil {
+						log.Println(err)
+					}
+					c.measureStorage.AddFavoriteNode(sub.Metadata.Ip, &storage.FavoriteNode{
+						IP:           sub.Metadata.Ip,
+						Address:      sub.Address,
+						Metadata:     base64.StdEncoding.EncodeToString(metadata),
+						Delay:        sub.Delay,
+						MinBandwidth: min / 1000,
+						MaxBandwidth: max / 1000,
+					})
+					err = c.measureStorage.SaveFavoriteNodes()
```

Comment:
I think better to let `AddFavoriteNode` return a bool indicating whether node is added or not, and only save if node is added